types/priv_validator.go

```diff
@@ -123,19 +113,43 @@ func LoadPrivValidator(filePath string) *PrivValidator {
 	return &privVal
 }
 
-func LoadOrGenPrivValidator(filePath string, logger log.Logger) *PrivValidator {
-	var privValidator *PrivValidator
-	if _, err := os.Stat(filePath); err == nil {
-		privValidator = LoadPrivValidator(filePath)
-		logger.Info("Loaded PrivValidator",
-			"file", filePath, "privValidator", privValidator)
-	} else {
-		privValidator = GenPrivValidator()
-		privValidator.SetFile(filePath)
-		privValidator.Save()
-		logger.Info("Generated PrivValidator", "file", filePath)
+// Generates a new validator with private key.
+func GenPrivValidator() *PrivValidator {
+	privKey := crypto.GenPrivKeyEd25519().Wrap()
+	pubKey := privKey.PubKey()
+	return &PrivValidator{
+		Address:  pubKey.Address(),
+		PubKey:   pubKey,
+		PrivKey:  privKey,
+		LastStep: stepNone,
+		filePath: "",
+		Signer:   NewDefaultSigner(privKey),
 	}
-	return privValidator
+}
+
+// LoadPrivValidatorWithSigner instantiates a private validator with a custom
+// signer object. Tendermint tracks state in the PrivValidator that might be
+// saved to disk. Please supply a filepath where Tendermint can save the
+// private validator.
+func LoadPrivValidatorWithSigner(signer Signer, filePath string) *PrivValidator {
+	return &PrivValidator{
+		Address:  signer.PubKey().Address(),
+		PubKey:   signer.PubKey(),
+		LastStep: stepNone,
+		filePath: filePath,
+		Signer:   signer,
+	}
+}
+
+func (privVal *PrivValidator) SetSigner(s Signer) {
+	privVal.Signer = s
+	privVal.setPubKeyAndAddress()
```

Comment:
rename to `updatePubKeyAndAddressFromSigner` or `derivePubKeyAndAddressFromSigner` ?